File: Example/STCardsMenu-ObjC/STCardsMenu-ObjC/STCardsMenu/STCardsMenuRootVC.m
//
//  STCardsMenuRootVC.m
//  STCardsMenu-ObjC
//
//  Created by TangJR on 12/14/15.
//  Copyright © 2015 tangjr. All rights reserved.
//

#import "STCardsMenuRootVC.h"
#import "UIColor+STCreation.h"
#import "STCardsMenuAnimation.h"
#import "STCardsMenuChildVC.h"
#import "STCardsMenuConst.h"

@interface STCardsMenuRootVC () <STCardsMenuChildVCDelegate, STCardsMenuClearWindowDelegate>

@property (assign, nonatomic) NSInteger childCount;
@property (assign, nonatomic) NSInteger lastSelectedIndex;
@property (strong, nonatomic) STCardsMenuCloseButton *closeButton;
@property (strong, nonatomic) STCardsMenuClearWindow *clearWindow;

@end

@implementation STCardsMenuRootVC

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.view.backgroundColor = [UIColor st_colorWithHex:0xCDDBED];
    [self closeButton];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.childCount = self.childViewControllers.count;
    for (NSInteger i = 0; i < self.childCount; i++) {
        STCardsMenuChildVC *childVC = self.childViewControllers[i];
        childVC.view.transform = CGAffineTransformMakeTranslation((i + 1) * STVCX, (i + 1) * STVCY);
        childVC.view.layer.cornerRadius = 8.0;
        childVC.delegate = self;
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self.clearWindow show];
}

- (void)clickAtIndex:(UInt8)selectIndex {
    if (selectIndex >= self.childCount) {
        return;
    }
    self.lastSelectedIndex = selectIndex;
    [self.clearWindow hide];
    [STCardsMenuAnimation closeButtonHideAnimationWithView:self.closeButton];
    STCardsMenuChildVC *selectVC = self.childViewControllers[selectIndex];
    [STCardsMenuAnimation hamburgShowAnimationWithView:selectVC.hamburgButton];
    [STCardsMenuAnimation titleMoveToRightAnimationWithView:selectVC.titleLabel];
    for (NSInteger i = 0; i <= selectIndex; i++) {
        STCardsMenuChildVC *childVC = self.childViewControllers[i];
        [STCardsMenuAnimation cardIdentityAnimationWithView:childVC.view];
    }
    for (NSInteger i = selectIndex + 1; i < self.childCount; i++) {
        STCardsMenuChildVC *childVC = self.childViewControllers[i];
        [STCardsMenuAnimation cardMoveToRightAnimationWithView:childVC.view];
    }
}

#pragma mark - STCardsMenuChildVCDelegate

- (void)cardsMenu:(STCardsMenuChildVC *)cardsMenuVC hamburgButtonTapped:(STCardsMenuHamburgButton *)hamburgButton {
    [self.clearWindow show];
    [STCardsMenuAnimation closeButtonShowAnimationWithView:self.closeButton];
    [STCardsMenuAnimation hamburgIdentityAnimationWithView:hamburgButton];
    [STCardsMenuAnimation titleIdentityAnimationWithView:cardsMenuVC.titleLabel];
    for (NSInteger i = 0; i < self.childCount; i++) {
        STCardsMenuChildVC *childVC = self.childViewControllers[i];
        [STCardsMenuAnimation cardScaleLittleAnimationWithView:childVC.view point:CGPointMake((i + 1) * STVCX, (i + 1) * STVCY)];
    }
}

- (void)cardsMenu:(STCardsMenuChildVC *)cardsMenuVC clearButtonTapped:(UIButton *)hamburgButton {
    NSInteger selectIndex = [self.childViewControllers indexOfObject:cardsMenuVC];
    [self clickAtIndex:selectIndex];
}

#pragma mark - STCardsMenuClearWindowDelegate

- (void)clearWindow:(STCardsMenuClearWindow *)window touchPoint:(CGPoint)touchPoint event:(UIEvent *)event {
    if (CGRectContainsPoint(self.closeButton.frame, touchPoint)) {
        [self clickAtIndex:self.lastSelectedIndex];
        return;
    }
    for (NSInteger i = self.childCount - 1; i >= 0; i--) {
        STCardsMenuChildVC *childVC = self.childViewControllers[i];
        BOOL isInside = CGRectContainsPoint(childVC.view.frame, touchPoint);
        if (isInside) {
            [self.clearWindow hide];
            [self clickAtIndex:i];
            break;
        }
    }
}

#pragma mark - Getter / Setter

- (STCardsMenuCloseButton *)closeButton {
    if (!_closeButton) {
        _closeButton = [STCardsMenuCloseButton new];
        _closeButton.translatesAutoresizingMaskIntoConstraints = NO;
        [self.view addSubview:_closeButton];
        NSLayoutConstraint *leftConstraint = [NSLayoutConstraint constraintWithItem:_closeButton attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self.view attribute:NSLayoutAttributeLeft multiplier:1 constant:STVCX];
        NSLayoutConstraint *topConstraint =  [NSLayoutConstraint constraintWithItem:_closeButton attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self.view attribute:NSLayoutAttributeTop multiplier:1 constant:STVCX];
        NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:_closeButton attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:30];
        NSLayoutConstraint *heightConstraint = [NSLayoutConstraint constraintWithItem:_closeButton attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:30];
        [self.view addConstraints:@[leftConstraint, topConstraint]];
        [_closeButton addConstraints:@[widthConstraint, heightConstraint]];
    }
    return _closeButton;
}

- (UIWindow *)clearWindow {
    if (!_clearWindow) {
        _clearWindow = [STCardsMenuClearWindow new];
        _clearWindow.delegate = self;
    }
    return _clearWindow;
}

@end

File: classes-Swift/STCardsMenu/STCardsMenuRootVC.swift
//
//  STCardsMenuRootVC.swift
//  STCardsMenu-Swift
//
//  Created by TangJR on 12/15/15.
//  Copyright © 2015 tangjr. All rights reserved.
//

import UIKit

class STCardsMenuRootVC: UIViewController {
    
    var childCount: Int = 0
    var lastSelectIndex: Int = 0
    let closeButton = STCardsMenuCloseButton()
    let clearWindow = STCardsMenuClearWindow()

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        view.backgroundColor = UIColor(st_hex: 0xCDDBED)
        setupCloseButton()
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        childCount = childViewControllers.count
        for i in 0 ..< childCount {
            let childVC = childViewControllers[i] as! STCardsMenuChildVC
            childVC.view.transform = CGAffineTransformMakeTranslation(CGFloat(i + 1) * STCardsMenuConst.VCX, CGFloat(i + 1) * STCardsMenuConst.VCY)
            childVC.view.layer.cornerRadius = 8.0
            childVC.delegate = self
        }
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        setupClearWindow()
    }
}

extension STCardsMenuRootVC: STCardsMenuChildVCDelegate {
    func hamburgButtonTapped(childVC: STCardsMenuChildVC, hamburgButton: STCardsMenuHamburgButton) {
        STCardsMenuAnimation.closeButtonShowAnimation(closeButton)
        STCardsMenuAnimation.hamburgButtonHideAnimation(hamburgButton, titleLabel: childVC.titleLabel)
        for i in 0 ..< childCount {
            let childVC = childViewControllers[i] as! STCardsMenuChildVC
            clearWindow.show()
            STCardsMenuAnimation.cardsMenuScaleLittleAnimation(childVC.view, point: CGPointMake((CGFloat(i) + 1.0) * STCardsMenuConst.VCX, (CGFloat(i) + 1.0) * STCardsMenuConst.VCY))
        }
    }
}

extension STCardsMenuRootVC: STCardsMenuClearWindowDelegate {
    func clearWindowTouched(clearWindow: STCardsMenuClearWindow, touchPoint: CGPoint) {
        if CGRectContainsPoint(closeButton.frame, touchPoint) {
            clickAtIndex(lastSelectIndex)
            return
        }
        for i in 0 ..< childCount {
            let childVC = childViewControllers[childCount - i - 1] as! STCardsMenuChildVC
            let isInside = CGRectContainsPoint(childVC.view.frame, touchPoint)
            if isInside {
                clickAtIndex(childCount - i - 1)
                break
            }
        }
    }
}

extension STCardsMenuRootVC {
    private func clickAtIndex(selectIndex: Int) {
        guard selectIndex <= childCount else {
            return
        }
        clearWindow.hide()
        lastSelectIndex = selectIndex
        let currentChildVC = childViewControllers[selectIndex] as! STCardsMenuChildVC
        STCardsMenuAnimation.hamburgButtonShowAnimation(currentChildVC.hamburgButton, titleLabel: currentChildVC.titleLabel)
        STCardsMenuAnimation.closeButtonHideAnimation(closeButton)
        for i in 0 ... selectIndex {
            let childVC = childViewControllers[i] as! STCardsMenuChildVC
            STCardsMenuAnimation.cardsMenuIdentityAnimation(childVC.view)
        }
        for i in selectIndex + 1 ..< childCount {
            let childVC = childViewControllers[i] as! STCardsMenuChildVC
            STCardsMenuAnimation.cardsMenuToRightAnimation(childVC.view)
        }
    }
    
    func closeButtonTapped(sender: UIButton) {
        clickAtIndex(lastSelectIndex)
    }
}

extension STCardsMenuRootVC {
    private func setupCloseButton() {
        view.addSubview(closeButton)
        closeButton.translatesAutoresizingMaskIntoConstraints = false
        let leftConstraint = NSLayoutConstraint(item: closeButton, attribute: .Left, relatedBy: .Equal, toItem: view, attribute: .Left, multiplier: 1.0, constant: STCardsMenuConst.VCX)
        let topConstraint = NSLayoutConstraint(item: closeButton, attribute: .Top, relatedBy: .Equal, toItem: view, attribute: .Top, multiplier: 1.0, constant: STCardsMenuConst.VCX)
        let widthConstraint = NSLayoutConstraint(item: closeButton, attribute: .Width, relatedBy: .Equal, toItem: nil, attribute: .NotAnAttribute, multiplier: 1.0, constant: 30.0)
        let heightConstraint = NSLayoutConstraint(item: closeButton, attribute: .Height, relatedBy: .Equal, toItem: nil, attribute: .NotAnAttribute, multiplier: 1.0, constant: 30.0)
        closeButton.addConstraints([widthConstraint, heightConstraint])
        view.addConstraints([leftConstraint, topConstraint])
    }
    
    private func setupClearWindow() {
        clearWindow.delegate = self
        clearWindow.show()
    }
}